tests/test_voting.py

```diff
@@ -0,0 +1,203 @@
+import pytest
+from collections import Counter
+
+from bigchaindb.core import Bigchain
+from bigchaindb.voting import Voting, INVALID, VALID, UNDECIDED
+
+
+################################################################################
+# Tests for checking vote eligibility
+
+
+def test_partition_eligible_votes():
+    class TestVoting(Voting):
+        @classmethod
+        def verify_vote_signature(cls, vote):
+            if vote['node_pubkey'] == 'invalid sig':
+                return False
+            if vote['node_pubkey'] == 'value error':
+                raise ValueError()
+            return True
+
+    voters = ['valid', 'invalid sig', 'value error', 'not in set']
+    votes = [{'node_pubkey': k} for k in voters]
+
+    el, inel = TestVoting.partition_eligible_votes(votes, voters[:-1])
+    assert el == [votes[0]]
+    assert inel == votes[1:]
+
+
+################################################################################
+# Test vote counting
+
+
+def test_count_votes():
+    class TestVoting(Voting):
+        @classmethod
+        def verify_vote_schema(cls, vote):
+            return vote['node_pubkey'] != 'malformed'
+
+    voters = (['cheat', 'cheat', 'says invalid', 'malformed'] +
+              ['kosher' + str(i) for i in range(10)])
+
+    votes = [Bigchain(v).vote('block', 'a', True) for v in voters]
+    votes[2]['vote']['is_block_valid'] = False
+    # Incorrect previous block subtracts from n_valid and adds to n_invalid
+    votes[-1]['vote']['previous_block'] = 'z'
+
+    assert TestVoting.count_votes(votes) == {
+        'counts': {
+            'n_valid': 9,    # 9 kosher votes
+            'n_invalid': 4,  # 1 cheat, 1 invalid, 1 malformed, 1 rogue prev block
+        },
+        'cheat': [votes[:2]],
+        'malformed': [votes[3]],
+        'previous_block': 'a',
+        'other_previous_block': {'z': 1},
+    }
+
+
+def test_must_agree_prev_block():
+    class TestVoting(Voting):
+        @classmethod
+        def verify_vote_schema(cls, vote):
+            return True
+
+    voters = 'abcd'
+    votes = [Bigchain(v).vote('block', 'a', True) for v in voters]
+    votes[0]['vote']['previous_block'] = 'b'
+    votes[1]['vote']['previous_block'] = 'c'
+    assert TestVoting.count_votes(votes) == {
+        'counts': {
+            'n_valid': 2,
+            'n_invalid': 2,
+        },
+        'previous_block': 'a',
+        'other_previous_block': {'b': 1, 'c': 1},
+        'malformed': [],
+        'cheat': [],
+    }
+
+
+################################################################################
+# Tests for vote decision making
+
+
+DECISION_TESTS = [dict(
```

Comment:
Sure; having just the table made it quite easy to look at the numbers and see what they meant though.